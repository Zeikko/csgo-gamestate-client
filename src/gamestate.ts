import * as t from 'io-ts'

const GameStateProviderCodec = t.type({
  name: t.string,
  appid: t.number,
  version: t.number,
  steamid: t.string,
  timestamp: t.number,
})
export type GameStateProvider = t.TypeOf<typeof GameStateProviderCodec>

const GameStateTeamCodec = t.type({
  score: t.number,
  name: t.string,
  flag: t.string,
  consecutive_round_losses: t.number,
  timeouts_remaining: t.number,
  matches_won_this_series: t.number,
})
export type GameStateTeam = t.TypeOf<typeof GameStateTeamCodec>

const GameStateModeCodec = t.union([t.literal('casual'), t.literal('competitive')])
export type GameStateMode = t.TypeOf<typeof GameStateModeCodec>

const GameStateMapPhaseCodec = t.union([
  t.literal('live'),
  t.literal('intermission'),
  t.literal('gameover'),
  t.literal('warmup'),
])
export type GameStateMapPhase = t.TypeOf<typeof GameStateMapPhaseCodec>

const GameStateMapCodec = t.type({
  mode: GameStateModeCodec,
  name: t.string,
  phase: GameStateMapPhaseCodec,
  round: t.number,
  team_ct: GameStateTeamCodec,
  team_t: GameStateTeamCodec,
  num_matches_to_win_series: t.number,
  current_spectators: t.number,
  souvenirs_total: t.number,
})
export type GameStateMap = t.TypeOf<typeof GameStateMapCodec>

const GameStateRoundPhaseCodec = t.union([t.literal('live'), t.literal('freezetime'), t.literal('over')])
export type GameStateRoundPhase = t.TypeOf<typeof GameStateRoundPhaseCodec>

const GameStateRoundType = t.type({
  phase: GameStateRoundPhaseCodec,
})
const GameStateRoundStatePartial = t.partial({
  bomb: t.string,
})
const GameStateRoundCodec = t.intersection([GameStateRoundType, GameStateRoundStatePartial])
export type GameStateRound = t.TypeOf<typeof GameStateRoundCodec>

const GameStateTeamStringCodec = t.union([t.literal('T'), t.literal('CT')])
export type GameStateTeamString = t.TypeOf<typeof GameStateTeamStringCodec>

const GameStateActivityCodec = t.union([
  t.literal('playing'),
  t.literal('free'),
  t.literal('textinput'),
  t.literal('menu'),
])
export type GameStateActivity = t.TypeOf<typeof GameStateActivityCodec>

const GameStatePlayerStateType = t.type({
  health: t.number,
  armor: t.number,
  helmet: t.boolean,
  flashed: t.number,
  burning: t.number,
  smoked: t.number,
  money: t.number,
  round_kills: t.number,
  round_killhs: t.number,
  round_totaldmg: t.number,
  equip_value: t.number,
})
const GameStatePlayerStatePartial = t.partial({
  defusekit: t.boolean,
})
const GameStatePlayerStateCodec = t.intersection([GameStatePlayerStateType, GameStatePlayerStatePartial])
export type GameStatePlayerState = t.TypeOf<typeof GameStatePlayerStateCodec>

const GameStatePlayerCodec = t.type({
  steamid: t.string,
  name: t.string,
  activity: GameStateActivityCodec,
  state: GameStatePlayerStateCodec,
})
export type GameStatePlayer = t.TypeOf<typeof GameStatePlayerCodec>

const GameStateAllPlayerStateType = t.type({
  health: t.number,
  armor: t.number,
  helmet: t.boolean,
  flashed: t.number,
  burning: t.number,
  money: t.number,
  round_kills: t.number,
  round_killhs: t.number,
  round_totaldmg: t.number,
  equip_value: t.number,
})
const GameStateAllPlayerStatePartial = t.partial({
  defusekit: t.boolean,
})
const GameStateAllPlayerStateCodec = t.intersection([GameStateAllPlayerStateType, GameStateAllPlayerStatePartial])
export type GameStateAllPlayerState = t.TypeOf<typeof GameStateAllPlayerStateCodec>

const GameStateAllPlayerMatchStatsCodec = t.type({
  kills: t.number,
  assists: t.number,
  deaths: t.number,
  mvps: t.number,
  score: t.number,
})
export type GameStateAllPlayerMatchStats = t.TypeOf<typeof GameStateAllPlayerMatchStatsCodec>

const GameStateWeaponStateCodec = t.union([t.literal('active'), t.literal('holstered')])
export type GameStateWeaponState = t.TypeOf<typeof GameStateWeaponStateCodec>

const GameStateWeaponType = t.type({
  name: t.string,
  paintkit: t.string,
  type: t.string,
  state: GameStateWeaponStateCodec,
})
const GameStateWeaponPartial = t.partial({
  ammo_clip: t.number,
  ammo_clip_max: t.number,
  ammo_reserve: t.number,
})
const GameStateWeaponCodec = t.intersection([GameStateWeaponType, GameStateWeaponPartial])
export type GameStateWeapon = t.TypeOf<typeof GameStateWeaponCodec>

const GameStateAllPlayerWeaponsCodec = t.partial({
  weapon_0: GameStateWeaponCodec,
  weapon_1: GameStateWeaponCodec,
  weapon_2: GameStateWeaponCodec,
  weapon_3: GameStateWeaponCodec,
  weapon_4: GameStateWeaponCodec,
  weapon_5: GameStateWeaponCodec,
  weapon_6: GameStateWeaponCodec,
})
export type GameStateAllPlayerWeapons = t.TypeOf<typeof GameStateAllPlayerWeaponsCodec>

const GameStateAllPlayerCodec = t.type({
  name: t.string,
  observer_slot: t.number,
  team: GameStateTeamStringCodec,
  state: GameStateAllPlayerStateCodec,
  match_stats: GameStateAllPlayerMatchStatsCodec,
  weapons: GameStateAllPlayerWeaponsCodec,
  position: t.string,
  forward: t.string,
})
export type GameStateAllPlayer = t.TypeOf<typeof GameStateAllPlayerCodec>

const GameStatePhaseCodec = t.union([
  t.literal('freezetime'),
  t.literal('bomb'),
  t.literal('live'),
  t.literal('over'),
  t.literal('defuse'),
  t.literal('paused'),
  t.literal('timeout_ct'),
  t.literal('timeout_t'),
])
export type GameStatePhase = t.TypeOf<typeof GameStatePhaseCodec>

const GameStatePhaseCountdownCodec = t.type({
  phase: GameStatePhaseCodec,
  phase_ends_in: t.string,
})
export type GameStatePhaseCountdown = t.TypeOf<typeof GameStatePhaseCountdownCodec>

const GameStateGrenadeNameCodec = t.union([
  t.literal('weapon_smokegrenade'),
  t.literal('weapon_decoy'),
  t.literal('weapon_hegrenade'),
  t.literal('weapon_incgrenade'),
  t.literal('weapon_molotov'),
])
export type GameStateGrenadeName = t.TypeOf<typeof GameStateGrenadeNameCodec>

const GameStateGrenadeType = t.type({
  owner: t.number,
  lifetime: t.string,
  type: t.string,
})
const GameStateGrenadePartial = t.partial({
  effecttime: t.string,
  position: t.string,
  velocity: t.string,
})
const GameStateGrenadeCodec = t.intersection([GameStateGrenadeType, GameStateGrenadePartial])
export type GameStateGrenade = t.TypeOf<typeof GameStateGrenadeCodec>

export const GameStateCodec = t.type({
  provider: GameStateProviderCodec,
  map: GameStateMapCodec,
  round: GameStateRoundCodec,
  player: GameStatePlayerCodec,
  allplayers: t.record(t.string, GameStateAllPlayerCodec),
  phase_countdowns: GameStatePhaseCountdownCodec,
  grenades: t.record(t.string, GameStateGrenadeCodec),
})
export type GameState = t.TypeOf<typeof GameStateCodec>
