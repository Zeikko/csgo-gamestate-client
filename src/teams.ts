import { GameState, GameStateTeam, GameStateTeamString } from "./gamestate"
import { improvePlayer, Player } from "./player"

export interface Teams {
  leftPlayers: Player[]
  rightPlayers: Player[]
  leftTeam?: Team
  rightTeam?: Team
}

export interface Team extends GameStateTeam {
  team: GameStateTeamString
}

export function getTeams(gameState: GameState): Teams {
  let leftPlayers
  let rightPlayers
  let leftTeam
  let rightTeam
  const allPlayers = Object.values(gameState.allplayers)
  const improvedPlayers = allPlayers.map(improvePlayer)
  const CTPlayers = improvedPlayers.filter((player) => player.team === 'CT')
  const TPlayers = improvedPlayers.filter((player) => player.team === 'T')
  const CTTeam = gameState.map.team_ct
  const TTeam = gameState.map.team_t
  const playerWithObserverSlot1 = improvedPlayers.find((player) => player.observerSlot === 1)
  if (!playerWithObserverSlot1) {
    return {
      leftPlayers: [],
      rightPlayers: [],
      leftTeam: undefined,
      rightTeam: undefined,
    }
  }
  const leftTeamSide = playerWithObserverSlot1.team
  if (leftTeamSide === 'CT') {
    leftPlayers = CTPlayers
    rightPlayers = TPlayers
    leftTeam = { ...CTTeam, team: 'CT' as GameStateTeamString }
    rightTeam = { ...TTeam, team: 'T' as GameStateTeamString }
  } else {
    leftPlayers = TPlayers
    rightPlayers = CTPlayers
    leftTeam = { ...TTeam, team: 'T' as GameStateTeamString }
    rightTeam = { ...CTTeam, team: 'CT' as GameStateTeamString }
  }
  return {
    leftPlayers,
    rightPlayers,
    leftTeam,
    rightTeam,
  }
}
