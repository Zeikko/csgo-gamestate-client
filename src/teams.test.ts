import { parseGameState } from './index'
import { getTeams } from './teams'
import { playingGameState } from './fixtures/gamestate'

describe('getTeams', () => {
  it('returns parse gamestate', () => {
    const parsed = parseGameState(playingGameState)
    expect(getTeams(parsed)).toMatchSnapshot()
  })
})
