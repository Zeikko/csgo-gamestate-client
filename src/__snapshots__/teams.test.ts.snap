// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getTeams returns parse gamestate 1`] = `
Object {
  "leftPlayers": Array [
    Object {
      "matchStats": Object {
        "assists": 3,
        "deaths": 19,
        "kills": 22,
        "mvps": 3,
        "score": 52,
      },
      "name": "allu",
      "observerSlot": 1,
      "state": Object {
        "armor": 100,
        "burning": 0,
        "defusekit": false,
        "equipmentValue": 6850,
        "flashed": 0,
        "health": 100,
        "helmet": true,
        "money": 500,
        "roundHeadshotsKills": 0,
        "roundKills": 3,
        "roundTotalDamage": 222,
      },
      "team": "T",
      "weapons": Object {
        "bomb": false,
        "grenades": Array [
          "weapon_molotov",
        ],
        "knife": "weapon_bayonet",
        "primary": "weapon_awp",
        "secondary": "weapon_glock",
      },
    },
    Object {
      "matchStats": Object {
        "assists": 6,
        "deaths": 23,
        "kills": 16,
        "mvps": 1,
        "score": 43,
      },
      "name": "sergej",
      "observerSlot": 2,
      "state": Object {
        "armor": 0,
        "burning": 0,
        "defusekit": false,
        "equipmentValue": 5000,
        "flashed": 0,
        "health": 0,
        "helmet": false,
        "money": 1550,
        "roundHeadshotsKills": 0,
        "roundKills": 0,
        "roundTotalDamage": 0,
      },
      "team": "T",
      "weapons": Object {
        "bomb": false,
        "grenades": Array [],
        "knife": undefined,
        "primary": undefined,
        "secondary": undefined,
      },
    },
    Object {
      "matchStats": Object {
        "assists": 3,
        "deaths": 21,
        "kills": 17,
        "mvps": 1,
        "score": 46,
      },
      "name": "Aleksib",
      "observerSlot": 3,
      "state": Object {
        "armor": 0,
        "burning": 0,
        "defusekit": false,
        "equipmentValue": 5000,
        "flashed": 0,
        "health": 0,
        "helmet": false,
        "money": 500,
        "roundHeadshotsKills": 0,
        "roundKills": 0,
        "roundTotalDamage": 0,
      },
      "team": "T",
      "weapons": Object {
        "bomb": false,
        "grenades": Array [],
        "knife": undefined,
        "primary": undefined,
        "secondary": undefined,
      },
    },
    Object {
      "matchStats": Object {
        "assists": 0,
        "deaths": 19,
        "kills": 18,
        "mvps": 3,
        "score": 45,
      },
      "name": "Aerial",
      "observerSlot": 4,
      "state": Object {
        "armor": 0,
        "burning": 0,
        "defusekit": false,
        "equipmentValue": 5000,
        "flashed": 0,
        "health": 0,
        "helmet": false,
        "money": 350,
        "roundHeadshotsKills": 0,
        "roundKills": 0,
        "roundTotalDamage": 0,
      },
      "team": "T",
      "weapons": Object {
        "bomb": false,
        "grenades": Array [],
        "knife": undefined,
        "primary": undefined,
        "secondary": undefined,
      },
    },
    Object {
      "matchStats": Object {
        "assists": 6,
        "deaths": 17,
        "kills": 21,
        "mvps": 7,
        "score": 71,
      },
      "name": "xseveN",
      "observerSlot": 5,
      "state": Object {
        "armor": 0,
        "burning": 0,
        "defusekit": false,
        "equipmentValue": 5000,
        "flashed": 0,
        "health": 0,
        "helmet": false,
        "money": 1250,
        "roundHeadshotsKills": 1,
        "roundKills": 1,
        "roundTotalDamage": 178,
      },
      "team": "T",
      "weapons": Object {
        "bomb": false,
        "grenades": Array [],
        "knife": undefined,
        "primary": undefined,
        "secondary": undefined,
      },
    },
  ],
  "leftTeam": Object {
    "consecutive_round_losses": 0,
    "flag": "FI",
    "matches_won_this_series": 0,
    "name": "ENCE",
    "score": 15,
    "team": "T",
    "timeouts_remaining": 1,
  },
  "rightPlayers": Array [
    Object {
      "matchStats": Object {
        "assists": 2,
        "deaths": 17,
        "kills": 21,
        "mvps": 3,
        "score": 55,
      },
      "name": "Jame",
      "observerSlot": 6,
      "state": Object {
        "armor": 0,
        "burning": 0,
        "defusekit": false,
        "equipmentValue": 7950,
        "flashed": 0,
        "health": 0,
        "helmet": false,
        "money": 2150,
        "roundHeadshotsKills": 0,
        "roundKills": 1,
        "roundTotalDamage": 100,
      },
      "team": "CT",
      "weapons": Object {
        "bomb": false,
        "grenades": Array [],
        "knife": undefined,
        "primary": undefined,
        "secondary": undefined,
      },
    },
    Object {
      "matchStats": Object {
        "assists": 5,
        "deaths": 19,
        "kills": 19,
        "mvps": 3,
        "score": 52,
      },
      "name": "buster",
      "observerSlot": 7,
      "state": Object {
        "armor": 100,
        "burning": 0,
        "defusekit": true,
        "equipmentValue": 7750,
        "flashed": 0,
        "health": 100,
        "helmet": true,
        "money": 6250,
        "roundHeadshotsKills": 0,
        "roundKills": 1,
        "roundTotalDamage": 100,
      },
      "team": "CT",
      "weapons": Object {
        "bomb": false,
        "grenades": Array [
          "weapon_hegrenade",
        ],
        "knife": "weapon_knife_m9_bayonet",
        "primary": "weapon_awp",
        "secondary": "weapon_usp_silencer",
      },
    },
    Object {
      "matchStats": Object {
        "assists": 5,
        "deaths": 20,
        "kills": 16,
        "mvps": 3,
        "score": 44,
      },
      "name": "KrizzeN",
      "observerSlot": 8,
      "state": Object {
        "armor": 0,
        "burning": 0,
        "defusekit": false,
        "equipmentValue": 6150,
        "flashed": 0,
        "health": 0,
        "helmet": false,
        "money": 3400,
        "roundHeadshotsKills": 1,
        "roundKills": 2,
        "roundTotalDamage": 200,
      },
      "team": "CT",
      "weapons": Object {
        "bomb": false,
        "grenades": Array [],
        "knife": undefined,
        "primary": undefined,
        "secondary": undefined,
      },
    },
    Object {
      "matchStats": Object {
        "assists": 5,
        "deaths": 19,
        "kills": 20,
        "mvps": 3,
        "score": 56,
      },
      "name": "fitch",
      "observerSlot": 9,
      "state": Object {
        "armor": 0,
        "burning": 0,
        "defusekit": false,
        "equipmentValue": 6150,
        "flashed": 0,
        "health": 0,
        "helmet": false,
        "money": 2800,
        "roundHeadshotsKills": 0,
        "roundKills": 0,
        "roundTotalDamage": 0,
      },
      "team": "CT",
      "weapons": Object {
        "bomb": false,
        "grenades": Array [],
        "knife": undefined,
        "primary": undefined,
        "secondary": undefined,
      },
    },
    Object {
      "matchStats": Object {
        "assists": 1,
        "deaths": 19,
        "kills": 23,
        "mvps": 0,
        "score": 53,
      },
      "name": "qikert",
      "observerSlot": 0,
      "state": Object {
        "armor": 0,
        "burning": 0,
        "defusekit": false,
        "equipmentValue": 5300,
        "flashed": 0,
        "health": 0,
        "helmet": false,
        "money": 7250,
        "roundHeadshotsKills": 0,
        "roundKills": 0,
        "roundTotalDamage": 0,
      },
      "team": "CT",
      "weapons": Object {
        "bomb": false,
        "grenades": Array [],
        "knife": undefined,
        "primary": undefined,
        "secondary": undefined,
      },
    },
  ],
  "rightTeam": Object {
    "consecutive_round_losses": 0,
    "flag": "KZ",
    "matches_won_this_series": 0,
    "name": "Avangar",
    "score": 12,
    "team": "CT",
    "timeouts_remaining": 3,
  },
}
`;
